src/redux/reducers.js

```diff
@@ -1,29 +1,53 @@
 // Initial State
+
 const initialState = {
-    count: 0,
-    updated: null,
-    limit: 1
+    habits: {}
 };
+/* habit structure
+const initialState = {
+    habits : {1: { id: 1, title: 'My habit', count: 0,limit: 1, log: [{updated: 1234567, info: { type: 'reset'}}]}}
+};*/
 
 const countReducer = (state = initialState, action) => {
     switch (action.type) {
-        case 'INCREMENT_COUNT': {
+        case 'CREATE_HABIT': {
+            const id = state.habits ? Object.keys(state.habits).length + 1 : 1;
+            const title = action.payload.title;
+            const limit = action.payload.limit;
+
             return {
                 ...state,
-                count: state.count + 1,
-                updated: action.payload.updated
+                habits: {
+                    ...state.habits,
+                    [id]: { id: id, title: title, count: 0, limit: limit, log: [{ updated: action.payload.updated, info: { type: 'created' } }] }
+                }
+            }
+        }
+        case 'INCREMENT_COUNT': {
+            const id = action.payload.id;
+            const habit = state.habits[id];
+            habit.count += 1;
+            habit.log.push({ updated: action.payload.updated, info: { type: 'increment' } });
+            return {
+                ...state
             }
         }
         case 'SET_LIMIT': {
+            const id = action.payload.id;
+            const habit = state.habits[id];
+            habit.limit = action.payload.limit;
+            habit.log.push({ updated: action.payload.updated, info: { type: 'limit' } });
             return {
-                ...state,
-                limit: action.payload.limit
+                ...state
             }
         }
         case 'RESET_COUNT': {
+            const id = action.payload.id;
+            const habit = state.habits[id];
+            habit.count = 0;
+            habit.log.push({ updated: action.payload.updated, info: { type: 'reset' } });
```

Comment:
I think these actions are making modifications to the original state value. You do use the spread operator in the return which will create a clone of the state rather than the same state object - but the original state object is updated with the updated habit details before this.

The spread operator makes a shallow clone, so when you spread an object the values are still pointing to the initial object's value. Here's an example of what I mean that I just ran in the `node` console:

```js
// Setup an initial state
state = { habit: { 0: { name: "Take meds" }, 1: { name: "Exercise" } } }

// Modify a habit
state.habit[0].name = "New name"

// Spread the initial state into a new state object - the new habit name is in place
newState = { ...state }
// > { habit: { '0': { name: 'New name' }, '1': { name: 'Exercise' } } }

// Inspect the original state - the new name is also in place
state
// > { habit: { '0': { name: 'New name' }, '1': { name: 'Exercise' } } }
```

Redux needs to be careful about this for a few reasons, simplest of which is that it means the library can no longer compare the old state to the new state in order to see which components will need to re-render.

- [Reducers and Immutable Updates](https://redux.js.org/tutorials/fundamentals/part-3-state-actions-reducers#reducers-and-immutable-updates)
- [Immutable Update Patterns](https://redux.js.org/usage/structuring-reducers/immutable-update-patterns#simplifying-immutable-updates-with-redux-toolkit)

Reply:
Updated reducers to prevent mutating state. I've followed the documentation and think I've implemented this correctly now. Might be worth casting your eye over again just in case...